### PR TITLE
Allow vmapping all_to_all and implement a (slow) CPU and GPU translation

### LINKED ISSUE #4141: No batching rule for `all_to_all`
```py
import jax
from jax import lax, vmap, pmap
import jax.numpy as jnp

def f(x):
  tiled_x = jax.tree_map(lambda x: jnp.tile(x[None], [jax.device_count()] + [1] * len(x.shape)), x)
  all_x = lax.pswapaxes(tiled_x, 'batch', 0)
  return all_x
pmap(vmap(f), axis_name='batch')(jnp.arange(jax.device_count() * 5 * 2).reshape((jax.device_count(), 5, 2)))
```

Currently this throws an assertion error elsewhere but with #4140, it will correctly throw `NotImplementedError: Batching rule for 'all_to_all' not implemented`.

### PR DESCRIPTION
**Pull request by apaszke**
This allows pmapping vmapped computations that use `all_to_all` or
`pswapaxes` inside. It also includes a very slow CPU and GPU translation
rule that might be useful for debugging programs locally, since XLA only
implements the `AllToAll` collective on TPUs.

Fixes #4141.

### CODE REVIEW

File: jax/lax/lax_parallel.py [hunk 1]
```diff
@@ -531,11 +546,23 @@ def _moveaxis(src, dst, x):
...
 
+def _all_to_all_batcher(vals_in, dims_in, *, axis_name, split_axis, concat_axis):
+  x, = vals_in
+  d, = dims_in
+  if split_axis <= d:
+    split_axis += 1
+  if concat_axis <= d:
+    concat_axis += 1
+  result = all_to_all_p.bind(x, axis_name=axis_name, split_axis=split_axis, concat_axis=concat_axis)
+  return result, d
```
Comment (apaszke) [inline jax/lax/lax_parallel.py:557]: Hm I just realized that I might need to adjust `d` here...

File: jax/lax/lax_parallel.py [hunk 2]
```diff
@@ -497,12 +496,28 @@ def _ppermute_batcher(vals_in, dims_in, axis_size, axis_name, perm):
 batching.collective_rules[ppermute_p] = _ppermute_batcher
 
 
+def _all_to_all_via_psum(x, *, axis_name, split_axis, concat_axis):
+  zero_full = lax.full((x.shape[split_axis],) + x.shape, 0, dtype=x.dtype)
+  idx = axis_index(axis_name)
+  local_full = lax.dynamic_update_slice_in_dim(zero_full, x[None], idx, 0)
+  global_full = psum(local_full, axis_name=axis_name)
```
Comment (mattjj) [inline jax/lax/lax_parallel.py:503]: I'm not sure, but can you call `_allgather` here? It seems similar.

File: jax/lax/lax_parallel.py (hunk 1 above)
Comment (mattjj) [reply jax/lax/lax_parallel.py:557]: Maybe try to add a test that fails?

**Review by mattjj — APPROVED** [review]
LGTM, modulo the batching rule you're looking at.

File: jax/lax/lax_parallel.py (hunk 1 above)
Comment (Thenerdstation) [reply jax/lax/lax_parallel.py:557]: I think it should be 

```python
# Vmapped axis got shifted down one.
if split_axis < concat_axis: 
  if split_axis <= d and concat_axis >= d:
    d -= 1
# Vmapped axis got shifted up one. 
if split_axis > concat_axis:
  if split_axis >= d and concat_axis <= d:
    d += 1
# In the case where the are equal, leave d alone.
```

**Comment by Thenerdstation** [comment]
Bump on this.

File: jax/lax/lax_parallel.py (hunk 2 above)
Comment (apaszke) [reply jax/lax/lax_parallel.py]: Nice catch, I used `all_gather` in the latest implementation

**Pull request merged**